tripy/tripy/dtype_info.py

```diff
@@ -16,35 +16,60 @@
+    count: int = None
 
-def dtype_info(dtype_variables: dict, dtype_constraints: dict):
-
+def dtype_info(
+    dtype_variables: dict = {},
+    dtype_constraints: dict = {},
+    param_type_specification: dict = {},
+    function_name : Optional[str] = "",
+):
+    """
+    This function is a decorator that populates TYPE_VERIFICATION global dictionary which will be used by
+    test_dtype_constraints.py to verify the dtypes of all operations.
+    """
     def decorator(func_obj):
-
+        # Get names and type hints for each param.
         func_sig = inspect.signature(func_obj)
         param_dict = func_sig.parameters
-        return_annotation = func_sig.return_annotation
-        # construct inputs_dict
+        # Construct inputs_dict.
         inputs_dict = {}
         for param_name, param_type in param_dict.items():
+            input_values = InputValues()
+            # If parameter had a default then use it otherwise skip.
+            if param_type.default is not param_type.empty:
+                # Checking if not equal to None since we want to enter if default is 0 or similar.
+                if param_type.default != None:
+                    input_values.init = param_type.default
             param_type = param_type.annotation
-            param_constraint_dict = {}
-            if param_type.startswith("tripy."):
-                param_type = param_type[6:]
-            param_constraint_dict[param_type] = {"dtype": dtype_constraints[param_name]}
-            inputs_dict[param_name] = param_constraint_dict
-        # construct returns_dict:
-        returns_dict = {}
-        if return_annotation.startswith("tripy."):
-            return_annotation = return_annotation[6:]
-        returns_dict[return_annotation] = {"dtype": dtype_constraints[RETURN_VALUE]}
-
-        parsed_dict = {"inputs": inputs_dict, "returns": returns_dict, "types": dtype_variables}
-        TYPE_VERIFICATION[func_obj.__qualname__] = (func_obj, parsed_dict, dtype_constraints)
+            # Check if there is a specific type that should be used.
+            if param_type_specification.get(param_name, None):
+                param_type = param_type_specification[param_name]
+            # If type is an optional or union get the first type.
+            while get_origin(param_type) in [Union, Optional]:
+                param_type = get_args(param_type)[0]
+                # ForwardRef refers to any case where type hint is a string.
+                if isinstance(param_type, ForwardRef):
+                    param_type = param_type.__forward_arg__
+            # Check if there are any provided constraints and add them to the constraint dict.
+            input_values.dtype = dtype_constraints.get(param_name, None)
+            inputs_dict[param_name] = {param_type: input_values}
+        return_dtype = dtype_constraints.get(RETURN_VALUE, -1)
+        parsed_dict = {"inputs": inputs_dict, "return_dtype": return_dtype, "types": dtype_variables}
```

Comment:
Can we rename `parsed_dict` to something more descriptive? What does it actually contain? Is it the type annotations?

Maybe a dataclass would be better here.

Reply:
I can just remove parsed dict completely and just add more elements to the tuples within `TYPE_VERIFICATION`.

Reply:
That makes sense. Could you make that tuple either a dataclass or NamedTuple?